common/setups/serialization.py

```diff
@@ -112,6 +126,72 @@ def _sis_hash(self):
         return sis_hash_helper(self.code_object)
 
 
+class PartialImport(Import):
+    """
+    Like Import, but for partial callables where certain parameters are given fixed and are hashed.
```

Comment:
> where certain parameters are given fixed 

Well.. this is not how `partial` works. Afaik the parameters that are given (and hashed) here are not fixed and can be overridden at runtime. Then we possibly have parameters influencing the hash that are not actually used or the hash changes without any different behavior (because the changed `hashed_arguments` from here are overwritten when actually calling the partial function/object.

Would this be a concern for us?

Reply:
Does partial not thrown an error if a parameter is given both "fixed" and "dynamically"? This would indeed be a problem.

Reply:
Then I am out of ideas.

Reply:
I didn't fully followed your discussion, but what exactly is the problem actually? It makes sense to me that `partial` behaves this way.

It's also very easy to have things influencing the hash, which would have no effect. E.g. put some random stuff into a `ReturnnConfig`. You cannot really avoid that the user does that. I would say it's up to the user.

It's also very easy to have things not influencing the hash, which would have an effect. Again, up to the user.

The more I work with Sisyphus, the more the feeling I get that 90% of our work is just because we care about the hashing logic so much, and quite often, it's wrong in the end anyway. And in a lot of cases, we are not fixing things now, or not cleaning things up, because it would change the hash, and that leads to a lot of messy legacy code just because of that... (@mmz33 can tell a few stories...)

Anyway, I don't want to get off topic here. My point is just that we cannot avoid such problems, and I don't really see a problem here.

Reply:
> Anyway, I don't want to get off topic here.

I like going off-topic here for a second, this can be used for tomorrow :) 

> The more I work with Sisyphus, the more the feeling I get that 90% of our work is just because we care about the hashing logic so much, and quite often, it's wrong in the end anyway

This only began to happen when we started sharing code. In my first 4 years of Sisyphus, everybody using it in MT had their own recipes, shared only what was needed and modified it to their own ideas of how their individual hashing should work. Of course there were tons of errors, but then it were your own errors and you would know about that sooner or later.  Still, Sisyphus allowed us to do experiments no other group was able to do easily at that time (looking at Miguels iterative backtranslation), and I was able to re-run a full setup for JTP some years after creating it and deleting all the data, by just activating the Sisyphus manager once.

Now when we share things, and all own setups are created in a way that you can actually update your recipes with other peoples code, there is no good solution to fixing/handling errros. RETURNN suffers a little (definitely not as much as here) from the same problem, people used wrong behavior (e.g. Batchnorm, ZoneoutLSTM) and now you can not remove it easily because setups depend on that. For i6_core we took over a lot of years-old untested and possibly buggy code just because there was no time to write it all new.

Nowadays I wonder if hasheable/controlled reproducible and "perfectly" shareable pipelines are even desired, because the needed engineering work distracts so much from actual research. If you get your work done, and the code avoids the biggest errors, we should consider good enough (different viewpoints may apply for I6 and AppTek). 

Long story short: multiple people are already using this (or previous versions) of this code to run successful experiments, so the interest in discussing this PR is understandably limited. So I agree with Alberts statement: There is no direct problem here. I think any further "safeguards" for the user would take unnecessary time, effort and increase code complexity.

Reply:
-- Off Topic Rant --
I remember my own time at the start of using sisyphus when every time you updated the recipes there was a solid chance that the manager would tell you either "Error in line XXX" when some interface was broken or "Jobs Runable: 10k" when some hash in the beginning of your pipeline changed and all your jobs would now be rerun. That was unpleasant and lead to people not pulling frequently which led to even higher chances of everything breaking when you finally needed a new feature. I do not want to go back there.

Then we decided to preserve the old behavior. That was good, because you could update now without loosing halve a day debugging the changes. But it was bad because we carried around a lot of broken legacy behavior. Then we decided to once in a year do a hash breaking change where all bugs would be fixed in a branch and then once you start a new set of experiments you switch over.
Well, Markus started a new set of experiments that moth, everyone else did not. Then we had a diverged set of branches and nobody could do adaptation. That was also not the solution.

With the switch to i6_core most of the (hash + compatibility) problems were resolved. Now even we can collaborate and use each others setups and `import_work_directory` is actually usable. (Tried that recently with a setup from Mohammad)
I really prefer this way and would never like to go back to the old days.

-- On topic --
> but what exactly is the problem actually?

No problem, actually. Just from the doc string I had the impression that @JackTemaki had a misunderstanding how partial works, which was confirmed later in this thread. We have clarified this misunderstanding and with the full knowledge we can answer the question that I posed:

> Would this be a concern for us?

and "No, this is the users problem" is a very fine answer that I do in fact agree with.
I might not use this class, but others are very welcome to if they want.

-- Off Topic Rant 2 --
I am a bit burned with partials, as I spent a significant amount of a day debugging an annoying problem that involved partials in combination with sets as default arguments - and then trying to find a solution that is backwards compatible.

Reply:
> here was a solid chance that the manager would tell you
Oh yes this was horrible, but I compared only the extremes to each other, having independent recipes compared to fully sharing them. Sharing them without keeping behavior was a nightmare.

> Now even we can collaborate and use each others setups
Did the same with my Hiwis, huge benefit.

> I really prefer this way and would never like to go back to the old days.
I do as well, this is not what I wanted to say. I just wanted to say it can never be "perfect". You can not have all benefits of constantly updating your code and keeping it clean together with full cooperation and stability. So we have to live with hash bugs and keeping old code as long as we are still more happy with the overall solution ;) 

> I might not use this class, but others are very welcome to if they want.
So you prefer the very explicit `PyTorchModel` solution?